src/Operation/FindOneAndReplace.php

```diff
@@ -147,10 +158,17 @@ public function __construct(string $databaseName, string $collectionName, $filte
 
         unset($options['projection'], $options['returnDocument']);
 
+        $replacementDocument = isset($options['codec'])
+            ? $options['codec']->encodeIfSupported($replacement)
+            : $replacement;
+        // Psalm's assert-if-true annotation does not work with unions, so
+        // assert the type manually instead of using is_document
+        assert(is_array($replacementDocument) || is_object($replacementDocument));
```

Comment:
Consider the following:

```php
if (isset($options['codec'])) {
    $replacement = $options['codec']->encodeIfSupported($replacement);
}

// Psalm's assert-if-true annotation does not work with unions, so
// assert the type manually instead of using is_document
assert(is_array($replacement) || is_object($replacement));
```

You could also conditionally move the `assert()` into the condition if that's all that's needed to satisfy Psalm. I would imagine so, as we already check `$replacement` earlier in the method.

----

I'm suggesting this because we don't really use `$replacement` for anything else, so I didn't see the harm in filtering/transforming the original value. This would apply in other contexts where you're modifying things, as well.

Perhaps not BulkWrite, as an explicitly named variable makes more sense there than `$args[i]`.